### ISSUE TITLE
Automatically map enum types using the enum datatype

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/sqlalchemy#8859]
Given an enum like
```py
class Foo(enum.Enum):
    foo = auto()
    bar = auto()
```

sqlalchemy could support mapping it directly with 

```py
class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    foo_col: Mapped[Foo]
```

Currently it needs to be mapped explicitly using type_annotation_map  or Annotated

Discussed in #8856

**Comment by zzzeek** [comment]
I also should be able to indicate the base type and overrides in type map, such as 

```py
class Base(DeclarativeBase):
    type_annotation_map = {
       enum.Enum: lambda python_type: sqlalchemy.Enum(python_type, native_enum=False),
       Foo: sqlalchemy.Enum(Foo, native_enum=True)
    }
```

note we need an indirection here such as the lambda example above in order to indicate options for the Enum mapping.   This implies the type_annotation_map would look for callables when it pulls out the type and resolves at mapping time if present.

**Comment by CaselIT** [comment]
ok, this is becoming complex real fast.

If we use mro of the enums only then we could have strange result happening, like

```py
class EnumMixin:
  def some_method(self):
    return 42

classs Foo(EnumMixin, enum.Enum):
    foo = auto()
    bar = auto()

class TypeMixin:
  def other_method(self):
    return 24

class MyType(Mixin, str):
  pass

class Base(DeclarativeBase):
    type_annotation_map = {
       EnumMixin: lambda python_type: sqlalchemy.Enum(python_type, native_enum=False),
       TypeMixin: sqlalchemy.Text()
    }
```

here enum mixin would be taken into consideration while type mixin would not work.

I'm starting to think that this will add a lot of complexity and wired cases and it's probably best to document this limitation and raise a proper error when detecting an enum on the left side

**Comment by zzzeek** [comment]
im not following exactly, did you write the correct names there?  TypeMixin doesnt seem related to the Enum classes there...

**Comment by CaselIT** [comment]
The strange thing would be that a mixin could be used for an enum, but not work for other types.
Seems kind of arbitrary.

As mentioned above, this seems something that can a lot of complexity for little gain

**Comment by zzzeek** [comment]
well it still makes sense if someone does `Mapped[MyEnum]`, that works, and it still makes sense that someone would want to configure the `Enum` type that gets created when someone says `Mapped[MyEnum]`.  So...suggestions?

**Comment by CaselIT** [comment]
detect the enum and raise an informative exception suggesting the use of annotated or type_map or set the type using mapped column?

**Comment by CaselIT** [comment]
enums seem to like creating issues and custom logic, see all the issues in alembic

**Comment by zzzeek** [comment]
alembic's issues are 95% because PostgreSQL has CREATE TYPE, which as a bonus can't be modified in a transaction.

**Comment by CaselIT** [comment]
(it does now (14+) but iirc you still can't use the new value until after the commit)

**Comment by zzzeek** [comment]
i figured they'd fix that at some point.  we wouldn't be able to ignore the issue for many years though (then again we haven't tried to fix enums for many years , so there's that!)

**Comment by sqla-tester** [comment]
**Federico Caselli** has proposed a fix for this issue in the **main** branch:

**Improve support for enum in mapped classes** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4249

**Comment by AlexanderPodorov** [comment]
Can we now fix Alembic's issue with dropping Enum type?

**Comment by CaselIT** [comment]
This issue is not related to dropping enum in alembic

**Comment by zzzeek** [comment]
this is way off topic but alembic's issue with PG ENUM is mostly that I don't have a clear API path to making all of that possible.  To the extent that PG's ENUM operations should be implicit vs. explicit in a migration file, where they come from, what they look like, how autogenerate would know what to do, then PG's *huge* restriction that changes to ENUMs can't be in a transaction (?!  i think they've recently fixed this though) all of that.   if people wanted to indicate what they'd like to see in migration files for all the use cases (not to mention we need to enumerate all the use cases), that would be a datapoint to start with.  

to date, nobody has made a single suggestion how they'd like to see things like modifying the elements in an ENUM to look from a public API perspective.

**Comment by CaselIT** [comment]
> then PG's huge restriction that changes to ENUMs can't be in a transaction (?! i think they've recently fixed this though)

not really, last time I looked (IIRC on v14) it supports adding a new value of an emun in a transaction, but that new value cannot be referenced anywhere until after a commit. Not sure if they changed something in v15
